File: web/src/store/action.ts
import { create } from "zustand";

import { Action } from "@/types/shared_types";

interface ActionState {
  selectedUser?: number;
  item: Action;
  list: Action[];

  setProperty: (key: keyof Action, value: any) => void;
  setList: (list: Action[]) => void;
  setUser: (userID: number) => void;
}

export const useActionStore = create<ActionState>((set) => ({
  item: {
    id: undefined,
    title: "",
    description: "",
    type: "",
    status: "",
    start_date: undefined,
    end_date: undefined,
    difficulty: 1,
    importance: 1,
    emotional_tax: 1,
  },
  list: [],
  selectedUser: undefined,

  setProperty: (property: keyof Action, value: any) => {
    set((state) => ({
      item: {
        ...state.item,
        [property]: value,
      },
    }));
  },
  setList: (list: Action[]) => {
    set(() => ({ list }));
  },
  setUser: (userID: number) => {
    set(() => ({ selectedUser: userID }));
  },
}));

File: api/internal/repositories/shared_types/action.go
package repositories

import (
	"time"
)

type Action struct {
	ID           int64     `json:"id,omitempty"`
	Title        string    `json:"title"`
	Description  string    `json:"description,omitempty"`
	Duration     int32     `json:"duration,omitempty"`
	Difficulty   int8      `json:"difficulty,omitempty"`
	Importance   int8      `json:"importance,omitempty"`
	EmotionalTax int8      `json:"emotional_tax,omitempty"`
	Priority     int8      `json:"priority,omitempty"`
	StartDate    time.Time `json:"start_date,omitempty"`
	EndDate      time.Time `json:"end_date,omitempty"`
	CreatedAt    time.Time `json:"create_at,omitempty"`
	UpdatedAt    time.Time `json:"updated_at,omitempty"`
}
